torch/csrc/jit/tensorexpr/tensorexpr_tutorical.py

```diff
@@ -0,0 +1,219 @@
+
+# 1. End-to-End example: addition of two tensors
+
+# We first create two BufHandles in TE which correspond to the two input
+# tensors.
+
+# The data (physical memory) of a tensor is represented as a buffer
+# in TE, which has dimensions, a data type and a name. BufHandle is a pointer
+# to a buffer that defines load/store operations on it. Here, A points to a
+# buffer with dimentions [32], 'float' data type and name 'A'; B points to a
+# buffer with dimentions [32], 'float' data type and name 'B'.
+A = te.BufHandle("A", [32], torch.float32)
+B = te.BufHandle("B", [32], torch.float32)
+
+# We then define the addition operation of two BufHandles: 'add'. It takes the
+# axis argument 'i' and returns the value of the ouput buffer on 'i', i.e.,
+# A[i] + B[i]. A[i], B[i] correspond to 'A.load([i])', 'B.load([i])' in TE
+# respectively.
+def add(i):
+    return A.load([i]) + B.load([i])
+
+# Next, we define the output tensor using TE 'Compute' - it specifies the name,
+# dimensions of the output tensor, and how to compute its values, i.e., the
+# 'add' operation we constructed above.
+C = te.Compute("C", [32], add)
+
+# At this point, we have constructed the IR stmt that computes the tensor C.
+# Use the following code to print its stmt.
+print("The stmt to compute tensor C:")
+print(C.stmt())
+
+# The stmt to compute tensor C:
+# for (int v = 0; v < 32; v++) {
+#   C[v] = (A[v]) + (B[v]);
+# }
+
+# Lastly, we construct a loopnest for C to prepare for codegen. Namely, we
+# transform the stmt to 1) expand reduce operations, and 2) flatten buffer
+# dimemsions. In this example, there are no reduces or multi-dimentional
+# buffers, thus the stmt remaining the same.
+loopnest = te.LoopNest([C])
+loopnest.prepare_for_codegen()
+
+# Print out the transformed stmt.
+print("The stmt before codegen:")
+print(loopnest.root_stmt())
+
+# The stmt before codegen:
+# {
+#   for (int v = 0; v < 32; v++) {
+#     C[v] = (A[v]) + (B[v]);
+#   }
+# }
+
+# Optional: simplify the stmt.
+# TODO: Should we add 'simplify' to 'prepare_for_codegen'? This seems to be a
+# common practise we do before codegen.
+stmt = te.simplify(loopnest.root_stmt())
+
+# Now we can get to codegen. Here, we use the IR Evaluator as the backend to
+# lower the stmt to assembly code. Besides the stmt, we also need to specify
+# the inputs and output data for codegen.
+cg = te.construct_codegen("ir_eval", stmt, [A, B, C])
+
+# We have generated native code in TE to add two tensors. Let's try it out!
+# Prepare the input/output tensors, and call TE kernel. The result is saved in
+# the output tensor `tC`.
+tA = torch.randn(32, dtype=torch.float32)
+tB = torch.randn(32, dtype=torch.float32)
+tC = torch.empty(32, dtype=torch.float32)
+cg.call_raw([tA.data_ptr(), tB.data_ptr(), tC.data_ptr()])
+
+# We can check the correctness of the results as follows.
+torch.testing.assert_close(tA + tB, tC)
+
+# Hoory, now you have learned how to create and run TE kernel!
+
+
+# 2. Construct your own TE IR
```

Comment:
It would be good to have a short intro in this chapter to explain what we're going to show and how it is connected to what we've done in the previous chapter.